osbenchmark/synthetic_data_generator/mapping_synthetic_data_generator.py

```diff
@@ -0,0 +1,438 @@
+        config_path (str): Path to the YAML file specifying data ranges.
+
+    Returns:
+        mapping and config as dicts
+    """
+    with open(mapping_file_path, "r") as f:
+        mapping_dict = json.load(f)
+
+    config_dict = None
+    if config_path:
+        with open(config_path, "r") as f:
+            config_dict = yaml.safe_load(f)
+
+    return mapping_dict, config_dict
+
+def generate_seeds_for_workers(regenerate=False):
+    client = get_client()
+    workers = client.scheduler_info()['workers']
+
+    seeds = []
+    for worker_id in workers.keys():
+        hash_object = hashlib.md5(worker_id.encode())
+
+        if regenerate:
+            # Add current timestamp to each hash to improve uniqueness
+            timestamp = str(time.time()).encode()
+            hash_object.update(timestamp)
+
+        hash_hex = hash_object.hexdigest()
+
+        seed = int(hash_hex[:8], 16)
+        seeds.append(seed)
+
+    return seeds
+
+def get_avg_document_size(index_mappings: dict, mapping_config: dict) -> int:
+    document = [generate_test_document(index_mappings, mapping_config)]
+    write_chunk(document, '/tmp/test-size.json')
+
+    size = os.path.getsize('/tmp/test-size.json')
+    os.remove('/tmp/test-size.json')
+
+    return size
+
+def generate_test_document(index_mappings: dict, mapping_config: dict) -> dict:
+    mapping_generator = MappingSyntheticDataGenerator(mapping_config)
+    converted_mappings = mapping_generator.transform_mapping_to_generators(index_mappings)
+
+    return mapping_generator.generate_fake_document(transformed_mapping=converted_mappings)
+
+def generate_dataset_with_mappings(client: Client, sdg_config: SyntheticDataGeneratorConfig, index_mappings: dict, input_config: dict):
+    """
+    param: client: Dask client that performs multiprocessing and creates dashboard to visualize task streams
+    param: sdg_config: SyntheticDataGenerationConfig instance that houses information related to data corpora to generate
+    param: index_mappings: OpenSearch index mappings that user provided
+    param: input_config: Optional config that specifies custom lists and custom data providers that the custom module uses to generate data.
+        This also contains configuration details related to how data is generated (i.e. number of workers to use, max file size in GB, and number of documents in a chunk)
+
+    returns: Does not return results but writes documents to output path
+    """
+    logger = logging.getLogger(__name__)
+
+    # Fetch settings and input config that user provided
+    generation_settings = get_generation_settings(input_config)
+    mapping_config = input_config
+
+    max_file_size_bytes = generation_settings.get('max_file_size_gb') * 1024 * 1024 * 1024
+    total_size_bytes = sdg_config.total_size_gb * 1024 * 1024 * 1024
+    chunk_size = generation_settings.get('chunk_size')
+    avg_document_size = get_avg_document_size(index_mappings, mapping_config)
+
+    current_size = 0
```

Comment:
Also this logic is duplicated in custom_synthetic_data_generator, having classes and using inheritance would definitely here reduce duplication.

Reply:
I initially implemented the custom mappings SDG and decided to quickly implement the mappings version but yes, I'll open up a subsequent PR to address this duplication.

Reply:
Will open a subsequent PR to address this in https://github.com/opensearch-project/opensearch-benchmark/issues/836